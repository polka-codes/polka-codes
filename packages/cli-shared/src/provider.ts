import { spawn } from 'node:child_process'
import { mkdir, readFile, rename, unlink, writeFile } from 'node:fs/promises'
import { dirname, normalize, resolve } from 'node:path'
import { input, select } from '@inquirer/prompts'
import type { ToolProvider } from '@polka-codes/core'
import ignore from 'ignore'
import { lookup } from 'mime-types'
import { checkRipgrep } from './utils/checkRipgrep'
import { listFiles } from './utils/listFiles'
import { searchFiles } from './utils/searchFiles'

export type ProviderOptions = {
  command?: {
    onStarted(command: string): void
    onStdout(data: string): void
    onStderr(data: string): void
    onExit(code: number): void
    onError(error: unknown): void
  }
  excludeFiles?: string[]
  summarizeOutput?: (stdout: string, stderr: string) => Promise<string | undefined>
  summaryThreshold?: number
}

export const getProvider = (options: ProviderOptions = {}): ToolProvider => {
  const ig = ignore().add(options.excludeFiles ?? [])
  const memoryStore: Record<string, string> = {}
  const provider: ToolProvider = {
    listTopics: async (): Promise<string[]> => {
      return Object.keys(memoryStore)
    },
    read: async (topic: string): Promise<string | undefined> => {
      return memoryStore[topic]
    },
    append: async (topic: string, content: string): Promise<void> => {
      if (memoryStore[topic]) {
        memoryStore[topic] += content
      } else {
        memoryStore[topic] = content
      }
    },
    replace: async (topic: string, content: string): Promise<void> => {
      memoryStore[topic] = content
    },
    remove: async (topic: string): Promise<void> => {
      delete memoryStore[topic]
    },
    readFile: async (path: string, includeIgnored: boolean): Promise<string | undefined> => {
      if (!includeIgnored && ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      try {
        return await readFile(path, 'utf8')
      } catch (_e) {
        return undefined
      }
    },
    writeFile: async (path: string, content: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      // generate parent directories if they don't exist
      await mkdir(dirname(path), { recursive: true })

      return await writeFile(path, content, 'utf8')
    },
    removeFile: async (path: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      return await unlink(path)
    },
    renameFile: async (sourcePath: string, targetPath: string): Promise<void> => {
      if (ig.ignores(sourcePath) || ig.ignores(targetPath)) {
        throw new Error(`Not allow to access file ${sourcePath} or ${targetPath}`)
      }
      return await rename(sourcePath, targetPath)
    },
    listFiles: async (path: string, recursive: boolean, maxCount: number, includeIgnored: boolean): Promise<[string[], boolean]> => {
      return await listFiles(path, recursive, maxCount, process.cwd(), options.excludeFiles, includeIgnored)
    },
    readBinaryFile: async (url: string) => {
      if (url.startsWith('file://')) {
        const filePath = decodeURIComponent(url.substring('file://'.length))
        const resolvedPath = normalize(resolve(process.cwd(), filePath))

        if (!resolvedPath.startsWith(process.cwd())) {
          throw new Error(`Access to file path "${filePath}" is restricted.`)
        }

        const data = await readFile(resolvedPath)
        const mediaType = lookup(resolvedPath) || 'application/octet-stream'

        return {
          base64Data: data.toString('base64'),
          mediaType,
        }
      }
      const response = await fetch(url)
      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`)
      }
      const data = await response.arrayBuffer()
      const mediaType = lookup(url) || 'application/octet-stream'

      return {
        base64Data: Buffer.from(data).toString('base64'),
        mediaType,
      }
    },

    executeCommand: (
      command: string,
      _needApprove: boolean,
    ): Promise<{ stdout: string; stderr: string; exitCode: number; summary?: string }> => {
      // TODO: add timeout

      return new Promise((resolve, reject) => {
        // spawn a shell to execute the command

        options.command?.onStarted(command)

        const child = spawn(command, [], {
          shell: true,
          stdio: 'pipe',
        })

        let stdoutText = ''
        let stderrText = ''

        child.stdout.on('data', (data) => {
          const dataStr = data.toString()
          options.command?.onStdout(dataStr)
          stdoutText += dataStr
        })

        child.stderr.on('data', (data) => {
          const dataStr = data.toString()
          options.command?.onStderr(dataStr)
          stderrText += dataStr
        })

        child.on('close', async (code) => {
          options.command?.onExit(code ?? 0)
          const totalLength = stdoutText.length + stderrText.length
          if (totalLength > (options.summaryThreshold ?? 5000) && options.summarizeOutput) {
            try {
              const summary = await options.summarizeOutput(stdoutText, stderrText)
              if (summary) {
                resolve({
                  summary,
                  stdout: stdoutText,
                  stderr: stderrText,
                  exitCode: code ?? 0,
                })
                return
              }
            } catch (_e) {
              console.error('Summarization failed:', _e)
            }
          }
          resolve({
            stdout: stdoutText,
            stderr: stderrText,
            exitCode: code ?? 0,
          })
        })

        child.on('error', (err) => {
          options.command?.onError(err)
          reject(err)
        })
      })
    },
    askFollowupQuestion: async (question: string, answerOptions: string[]): Promise<string> => {
      if (answerOptions.length === 0) {
        return await input({ message: question })
      }

      const otherMessage = 'Other (enter text)'
      answerOptions.push(otherMessage)
      const answer = await select({
        message: question,
        choices: answerOptions.map((option) => ({ name: option, value: option })),
      })

      if (answer === otherMessage) {
        return await input({ message: 'Enter your answer:' })
      }
      return answer
    },
    fetchUrl: async (url: string): Promise<string> => {
      const isRaw = url.startsWith('https://raw.githubusercontent.com/')

      const urlToFetch = isRaw ? url : `https://r.jina.ai/${url}`

      try {
        const response = await fetch(urlToFetch)
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`)
        }
        return await response.text()
      } catch (error) {
        console.error('Error fetching URL:', error)
        throw error
      }
    },
  }

  if (checkRipgrep()) {
    provider.searchFiles = async (path: string, regex: string, filePattern: string): Promise<string[]> => {
      return await searchFiles(path, regex, filePattern, process.cwd(), options.excludeFiles)
    }
  } else {
    console.error(
      'Error: ripgrep (rg) is not installed. Search file tool is disabled. Please install it from https://github.com/BurntSushi/ripgrep#installation',
    )
  }

  return provider
}
