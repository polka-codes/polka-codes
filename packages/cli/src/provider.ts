import { spawn } from 'node:child_process'
import { mkdir, readFile, writeFile } from 'node:fs/promises'
import { dirname } from 'node:path'
import ignore from 'ignore'

import type { ToolProvider } from '@polka-codes/core'
import { listFiles } from './utils/listFiles'
import { searchFiles } from './utils/searchFiles'

export type ProviderOptions = {
  command: {
    onStarted(command: string): void
    onStdout(data: string): void
    onStderr(data: string): void
    onExit(code: number): void
    onError(error: unknown): void
  }
  excludeFiles?: string[]
}

export const getProvider = (options: ProviderOptions): ToolProvider => {
  const ig = ignore().add(options.excludeFiles ?? [])
  return {
    readFile: async (path: string): Promise<string> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      return await readFile(path, 'utf8')
    },
    writeFile: async (path: string, content: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      // generate parent directories if they don't exist
      await mkdir(dirname(path), { recursive: true })

      return await writeFile(path, content, 'utf8')
    },
    listFiles: async (path: string, recursive: boolean, maxCount: number): Promise<[string[], boolean]> => {
      return await listFiles(path, recursive, maxCount, dirname(path), options.excludeFiles)
    },
    searchFiles: async (path: string, regex: string, filePattern: string): Promise<string[]> => {
      return await searchFiles(path, regex, filePattern, dirname(path), options.excludeFiles)
    },
    // listCodeDefinitionNames: async (path: string) => Promise<string[]> {},

    executeCommand: (command: string, needApprove: boolean): Promise<{ stdout: string; stderr: string; exitCode: number }> => {
      // TODO: add timeout

      return new Promise((resolve, reject) => {
        // spawn a shell to execute the command

        options.command.onStarted(command)

        const child = spawn(command, [], {
          shell: true,
          stdio: 'pipe',
        })

        let stdoutText = ''
        let stderrText = ''

        child.stdout.on('data', (data) => {
          const dataStr = data.toString()
          options.command.onStdout(dataStr)
          stdoutText += dataStr
        })

        child.stderr.on('data', (data) => {
          const dataStr = data.toString()
          options.command.onStderr(dataStr)
          stderrText += dataStr
        })

        child.on('close', (code) => {
          options.command.onExit(code ?? 0)
          resolve({
            stdout: stdoutText,
            stderr: stderrText,
            exitCode: code ?? 0,
          })
        })

        child.on('error', (err) => {
          options.command.onError(err)
          reject(err)
        })
      })
    },
    // askFollowupQuestion: async (question: string, options: string[]) => Promise<string> {},
    // attemptCompletion: async (result: string) => Promise<string | undefined> {},
  }
}
