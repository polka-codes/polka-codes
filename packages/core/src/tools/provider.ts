export type FilesystemProvider = {
  readFile?: (path: string) => Promise<string>
  writeFile?: (path: string, content: string) => Promise<void>
  removeFile?: (path: string) => Promise<void>
  listFiles?: (path: string, recursive: boolean, maxCount: number) => Promise<[string[], boolean]>
  searchFiles?: (path: string, regex: string, filePattern: string) => Promise<string[]>
  listCodeDefinitionNames?: (path: string) => Promise<string[]>
}

export type CommandProvider = {
  executeCommand?: (command: string, needApprove: boolean) => Promise<{ stdout: string; stderr: string; exitCode: number }>
}

export type InteractionProvider = {
  askFollowupQuestion?: (question: string, options: string[]) => Promise<string>
  attemptCompletion?: (result: string) => Promise<string | undefined>
}

export type ToolProvider = FilesystemProvider & CommandProvider & InteractionProvider

export class MockProvider implements ToolProvider {
  async readFile(path: string): Promise<string> {
    return 'mock content'
  }

  async writeFile(path: string, content: string): Promise<void> {
    return
  }

  async removeFile(path: string): Promise<void> {
    return
  }

  async listFiles(path: string, recursive: boolean, maxCount: number): Promise<[string[], boolean]> {
    return [['mock-file.txt'], false]
  }

  async searchFiles(path: string, regex: string, filePattern: string): Promise<string[]> {
    return ['mock-file.txt']
  }

  async listCodeDefinitionNames(path: string): Promise<string[]> {
    return ['mockDefinition']
  }

  async executeCommand(command: string, needApprove: boolean): Promise<{ stdout: string; stderr: string; exitCode: number }> {
    return { stdout: 'mock output', stderr: '', exitCode: 0 }
  }

  async askFollowupQuestion(question: string, options?: string[]): Promise<string> {
    return 'mock answer'
  }

  async attemptCompletion(result: string): Promise<string | undefined> {
    return 'mock completion'
  }
}
