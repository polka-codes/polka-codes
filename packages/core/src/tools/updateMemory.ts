import { z } from 'zod'
import { type FullToolInfo, type ToolHandler, type ToolInfo, ToolResponseType } from '../tool'
import type { MemoryProvider } from './provider'

export const toolInfo = {
  name: 'updateMemory',
  description: 'Appends, replaces, or removes content from a memory topic.',
  parameters: z
    .object({
      operation: z.enum(['append', 'replace', 'remove']).describe('The operation to perform.'),
      topic: z.string().nullish().describe('The topic to update in memory. Defaults to ":default:".'),
      content: z.string().optional().describe('The content for append or replace operations. Must be omitted for remove operation.'),
    })
    .superRefine((data, ctx) => {
      if (data.operation === 'append' || data.operation === 'replace') {
        if (data.content === undefined) {
          ctx.addIssue({
            code: 'custom',
            message: 'Content is required for "append" and "replace" operations.',
            path: ['content'],
          })
        }
      } else if (data.operation === 'remove') {
        if (data.content !== undefined) {
          ctx.addIssue({
            code: 'custom',
            message: 'Content must not be provided for "remove" operation.',
            path: ['content'],
          })
        }
      }
    }),
} as const satisfies ToolInfo

export const handler: ToolHandler<typeof toolInfo, MemoryProvider> = async (provider, args) => {
  if (!provider.updateMemory) {
    return {
      type: ToolResponseType.Error,
      message: {
        type: 'error-text',
        value: 'Memory operations are not supported by the current provider.',
      },
    }
  }
  const params = toolInfo.parameters.parse(args)

  await provider.updateMemory(params.operation, params.topic ?? undefined, 'content' in params ? params.content : undefined)

  switch (params.operation) {
    case 'append':
      return {
        type: ToolResponseType.Reply,
        message: {
          type: 'text',
          value: `Content appended to memory topic '${params.topic || ':default:'}'.`,
        },
      }
    case 'replace':
      return {
        type: ToolResponseType.Reply,
        message: {
          type: 'text',
          value: `Memory topic '${params.topic || ':default:'}' replaced.`,
        },
      }
    case 'remove':
      return {
        type: ToolResponseType.Reply,
        message: {
          type: 'text',
          value: `Memory topic '${params.topic || ':default:'}' removed.`,
        },
      }
  }
}

export default {
  ...toolInfo,
  handler,
} satisfies FullToolInfo
