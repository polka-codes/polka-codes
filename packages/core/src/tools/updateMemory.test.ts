import { describe, expect, it, spyOn } from 'bun:test'
import { MockProvider } from './provider'
import updateMemory from './updateMemory'

describe('updateMemory', () => {
  describe('parameters', () => {
    it('should parse append operation with content', () => {
      const result = updateMemory.parameters.safeParse({
        operation: 'append',
        content: 'new data',
      })
      expect(result.success).toBe(true)
    })

    it('should fail to parse append operation without content', () => {
      const result = updateMemory.parameters.safeParse({
        operation: 'append',
      })
      expect(result.success).toBe(false)
      expect(result.error).toMatchSnapshot()
    })

    it('should parse replace operation with content', () => {
      const result = updateMemory.parameters.safeParse({
        operation: 'replace',
        content: 'new data',
      })
      expect(result.success).toBe(true)
    })

    it('should fail to parse replace operation without content', () => {
      const result = updateMemory.parameters.safeParse({
        operation: 'replace',
      })
      expect(result.success).toBe(false)
      expect(result.error).toMatchSnapshot()
    })

    it('should parse remove operation without content', () => {
      const result = updateMemory.parameters.safeParse({
        operation: 'remove',
      })
      expect(result.success).toBe(true)
    })

    it('should fail to parse remove operation with content', () => {
      const result = updateMemory.parameters.safeParse({
        operation: 'remove',
        content: 'some data',
      })
      expect(result.success).toBe(false)
      expect(result.error).toMatchSnapshot()
    })
  })

  describe('handler', () => {
    it('should handle append operation', async () => {
      const mockProvider = new MockProvider()
      const spy = spyOn(mockProvider, 'updateMemory')

      const result = await updateMemory.handler(mockProvider, {
        operation: 'append',
        content: 'new data',
      })

      expect(result).toMatchSnapshot()
      expect(spy).toHaveBeenCalledWith('append', undefined, 'new data')
    })

    it('should handle replace operation with topic', async () => {
      const mockProvider = new MockProvider()
      const spy = spyOn(mockProvider, 'updateMemory')

      const result = await updateMemory.handler(mockProvider, {
        operation: 'replace',
        topic: 'my-topic',
        content: 'new data',
      })

      expect(result).toMatchSnapshot()
      expect(spy).toHaveBeenCalledWith('replace', 'my-topic', 'new data')
    })

    it('should handle remove operation', async () => {
      const mockProvider = new MockProvider()
      const spy = spyOn(mockProvider, 'updateMemory')

      const result = await updateMemory.handler(mockProvider, {
        operation: 'remove',
      })

      expect(result).toMatchSnapshot()
      expect(spy).toHaveBeenCalledWith('remove', undefined, undefined)
    })

    it('should return error if provider does not support memory', async () => {
      const mockProvider = new MockProvider()
      mockProvider.updateMemory = undefined as any

      const result = await updateMemory.handler(mockProvider, {
        operation: 'append',
        content: 'new data',
      })

      expect(result).toMatchSnapshot()
    })
  })
})
