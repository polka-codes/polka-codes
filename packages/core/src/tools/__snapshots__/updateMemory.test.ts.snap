// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`updateMemory parameters should fail to parse append operation without content 1`] = `[Definition: [
  {
    "code": "custom",
    "message": "Content is required for \\"append\\" and \\"replace\\" operations.",
    "path": [
      "content"
    ]
  }
]]`;

exports[`updateMemory parameters should fail to parse replace operation without content 1`] = `[Definition: [
  {
    "code": "custom",
    "message": "Content is required for \\"append\\" and \\"replace\\" operations.",
    "path": [
      "content"
    ]
  }
]]`;

exports[`updateMemory parameters should fail to parse remove operation with content 1`] = `[Definition: [
  {
    "code": "custom",
    "message": "Content must not be provided for \\"remove\\" operation.",
    "path": [
      "content"
    ]
  }
]]`;

exports[`updateMemory handler should handle append operation 1`] = `
{
  "message": {
    "type": "text",
    "value": "Content appended to memory topic ':default:'.",
  },
  "type": "Reply",
}
`;

exports[`updateMemory handler should handle replace operation with topic 1`] = `
{
  "message": {
    "type": "text",
    "value": "Memory topic 'my-topic' replaced.",
  },
  "type": "Reply",
}
`;

exports[`updateMemory handler should handle remove operation 1`] = `
{
  "message": {
    "type": "text",
    "value": "Memory topic ':default:' removed.",
  },
  "type": "Reply",
}
`;

exports[`updateMemory handler should return error if provider does not support memory 1`] = `
{
  "message": {
    "type": "error-text",
    "value": "Memory operations are not supported by the current provider.",
  },
  "type": "Error",
}
`;
